### PR TITLE
Initialize the block allocated by `caml_alloc_shr_check_gc`

### PR DESCRIPTION
**Pull request by xavierleroy**
This is the quick fix for issue #12481 mentioned in https://github.com/ocaml/ocaml/issues/12481#issuecomment-1682627736 .  I think it's appropriate as an emergency fix for the 5.1 release, even though we may want to consider #12485 or other approaches later.

Fixes: #12481

### CODE REVIEW

**Comment by gasche** [comment]
It took me a while to understand the correctness argument for this fix. (I am still not sure whether it can actually be the case that `clos_vars` contain complex expressions that call a GC on evaluation, or whether GC during field-filling can only happen with other sort of blocks.)

I think that it would be helpful to write down some of this correctness argument in the code, not just across tricky github issues. I wrote a comment on top of your PR in my branch:   https://github.com/gasche/ocaml/commit/emergency-fix-12481
Would you be wiling to integrate something like this with the fix, with whatever rephrasing you like?

**Comment by lthls** [comment]
> (I am still not sure whether it can actually be the case that `clos_vars` contain complex expressions that call a GC on evaluation, or whether GC during field-filling can only happen with other sort of blocks.)

It is uncommon for closures to contain complex expressions, but you can trigger that with unboxing:
```ocaml
let f x =
  let y = x +. 0. in
  fun z -> y +. z
```
Flambda can also push arbitrary expressions into closures during the `un_anf` pass, I believe.

**Comment by xavierleroy** [comment]
Thanks for the comment, I added it to this PR, along with another comment in runtime/alloc.c.  (Wow, so many comments!)

**Review by gasche — APPROVED** [review]
Looks good to me.

(We probably need a go-ahead of @Octachron before merging in 5.1, the target of this PR.)

**Comment by xavierleroy** [comment]
Thanks for the approval.  Yes, we will let @octachron decide what to do with this PR.

**Comment by Octachron** [comment]
I agree that it seems better to have this quick fix in 5.1.0 .

**Comment by kayceesrk** [comment]
Thanks for the quick fix @xavierleroy.

**Pull request merged**
